cmd/thanos/query_frontend.go

```diff
@@ -146,6 +147,10 @@ func registerQueryFrontend(app *extkingpin.App) {
 
 	cmd.Flag("query-frontend.forward-header", "List of headers forwarded by the query-frontend to downstream queriers, default is empty").PlaceHolder("<http-header-name>").StringsVar(&cfg.ForwardHeaders)
 
+	cmd.Flag("query-frontend.tenant-header", "HTTP header to determine tenant").Default(tenancy.DefaultTenantHeader).Hidden().StringVar(&cfg.TenantHeader)
```

Comment:
I'm not sure if this is the right place to discuss this, but the query frontend also has a `query-frontend.org-id-header` arg. Thanos side, I believe this is only used when logging slow queries, but probably has some more use for Cortex.

Do we want to keep both the `org-id-header` and the `tenant-header` long term ? Could perhaps be nice to unify that into one cmd line arg if possible.

Reply:
Long term we should indeed converge on only one to avoid confusion for users and ourselves, and for consistency. 

As part of the migration process, I can mark the `query-frontend.org-id-header` as deprecated for now, mention the new one, and store the configured `org-id-header` in the tenant header config. This way someone already using this will have a warning about the change and can easily adapt to the new CLI arg, allowing us to completely drop the org ID config in a future breaking change. 

Internally and more technically speaking, I am copying the tenant header field's value to the Org ID one because of some middlewares we are importing/reusing from Cortex, but I think over time we can adapt them to avoid this and clean things up.

What do you think, @jacobbaungard?

Reply:
Sounds like a good plan to me. What would happen if one define both, can we error out in that case?

Reply:
@jacobbaungard we could error out or print out a warning & give priority to the tenant header config, similar to what happens if you configure the header and the cert field in Receive. I personally have no preference. Will update to error out when both are provided. 👍

Reply:
Ok, I think the `orgID` was introduced only as a method of identifying sources of slow queries and passing that information to Cortex. It's also a repeatable flag, with significance in how headers are ordered, afaiu reading original [issue](https://github.com/thanos-io/thanos/issues/3177)/[PR](https://github.com/thanos-io/thanos/pull/3245).

Thus I don't think it is mutually exclusive with this `tenant` header, and both have different semantics as well, as you can have multiple `orgID` headers, but only a single `tenant` one. So let's not error or warn the user, and keep it for now.

cc: @yeya24 in case he has some other thoughts on this! 🙂